modules/solid_mechanics/include/neml2/userobjects/MOOSEToNEML2Batched.h

```diff
@@ -32,17 +46,82 @@ class MOOSEToNEML2Batched : public MOOSEToNEML2, public ElementUserObject
 
   neml2::Tensor gatheredData() const override;
 
-  // The number of gathered data items (for setting the model batch size)
-  std::size_t size() const { return _buffer.size(); }
+  // The number of elements
+  std::size_t nElem() const { return _buffer.size(); }
+
+  /// The number of quadrature points in each element
+  // Note that this assumes that all elements have the same number of quadrature points
+  std::size_t nQP() const { return _nqp; }
 
 protected:
-  /// Convert the underlying MOOSE data to a torch::Tensor
-  virtual torch::Tensor convertQpMOOSEData() const = 0;
+  /// MOOSE data for the current element
+  virtual const MooseArray<T> & elemMOOSEData() const = 0;
 
   /// Intermediate data buffer, filled during the element loop
-  std::vector<torch::Tensor> _buffer;
+  std::vector<MooseArray<T>> _buffer;
 
-  /// Current element's quadrature point indexing
-  unsigned int _qp;
+  /// Number of quadrature points
+  std::size_t _nqp;
 #endif
 };
+
+template <typename T>
+InputParameters
+MOOSEToNEML2Batched<T>::validParams()
+{
+  auto params = MOOSEToNEML2::validParams();
+  params += ElementUserObject::validParams();
+
+  // Since we use the NEML2 model to evaluate the residual AND the Jacobian at the same time, we
+  // want to execute this user object only at execute_on = LINEAR (i.e. during residual evaluation).
+  // The NONLINEAR exec flag below is for computing Jacobian during automatic scaling.
+  ExecFlagEnum execute_options = MooseUtils::getDefaultExecFlagEnum();
+  execute_options = {EXEC_INITIAL, EXEC_LINEAR, EXEC_NONLINEAR};
+  params.set<ExecFlagEnum>("execute_on") = execute_options;
+
+  return params;
+}
+
+template <typename T>
+MOOSEToNEML2Batched<T>::MOOSEToNEML2Batched(const InputParameters & params)
+  : MOOSEToNEML2(params), ElementUserObject(params)
+{
+}
+
+#ifdef NEML2_ENABLED
+template <typename T>
+void
+MOOSEToNEML2Batched<T>::initialize()
+{
+  _buffer.clear();
+}
+
+template <typename T>
+void
+MOOSEToNEML2Batched<T>::execute()
+{
+  if (_buffer.empty())
+    _nqp = elemMOOSEData().size();
+  else
+    mooseAssert(_nqp == elemMOOSEData().size(),
+                "Number of quadrature points must be the same for all elements");
+  _buffer.push_back(this->elemMOOSEData());
+}
+
+template <typename T>
+void
+MOOSEToNEML2Batched<T>::threadJoin(const UserObject & uo)
+{
+  // append vectors
+  const auto & m2n = static_cast<const MOOSEToNEML2Batched<T> &>(uo);
+  _buffer.insert(_buffer.end(), m2n._buffer.begin(), m2n._buffer.end());
+  mooseAssert(_nqp == m2n._nqp, "Number of quadrature points must be the same for all elements");
```

Comment:
That was my (unsuccessful) attempt at using memcpy to copy the entire neml2 tensor to moose material storage. I've reverted this change if you look at my updated branch. That is, the batch indexing is still linear, with element IDs as offsets and nqp as strides.

This attempt was unsuccessful also because face elements could have different number of qps than volume elements (even when block restricted).